src/string_bytes.cc

```diff
@@ -72,10 +73,15 @@ class ExternString: public ResourceType {
       ExternString* h_str = new ExternString<ResourceType, TypeName>(isolate,
                                                                      data,
                                                                      length);
-      Local<String> str = String::NewExternal(isolate, h_str);
+      MaybeLocal<String> str = String::NewExternal(isolate, h_str);
       isolate->AdjustAmountOfExternalAllocatedMemory(h_str->byte_length());
 
-      return scope.Escape(str);
+      if (str.IsEmpty()) {
+        delete h_str;
+        return Local<String>();
+      }
```

Comment:
Your example code does not use `NewFromCopy` because it uses the fast lane via `NewFromUtf8` and then returns `undefined` because v8 returns an empty handle if `length > kStringMaxLength`. 

Example code with `NewFromCopy`:
````
var a = [];
var buffer = new Buffer(200 * 1024 * 1024);
while (true) {
  a.push(buffer.toString('binary'));
}
````
If `new TypeName[length]` fails:
````
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
[1]    2013 abort (core dumped)  ./iojs test
````